### PR TITLE
table: initialize index.needRestoredData lazily

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/24029 

Problem Summary:

`NewIndex()` relies on `NeedRestoredData`, but it is not initialized in the bootstrap stage.

### What is changed and how it works?

What's Changed: Initialize index.needRestoredData lazily

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

NA

### Release note 

- Fix row & index inconsistent with new collation enabled after upgrade.

### CODE REVIEW

File: table/tables/index.go [hunk 1]
```diff
@@ -176,6 +180,9 @@ func (c *index) Create(sctx sessionctx.Context, txn kv.Transaction, indexedValue
 
 	// save the key buffer to reuse.
 	writeBufs.IndexKeyBuf = key
+	c.initNeedRestoreData.Do(func() {
+		c.needRestoredData = NeedRestoredData(c.idxInfo.Columns, c.tblInfo.Columns)
```
Comment (coocood) [inline table/tables/index.go:184]: We may need to create index during upgrade in the future.

Comment (tangenta) [reply table/tables/index.go:184]: If the index creator has no idea about whether `new collation` is enabled, how can it create?

Comment (coocood) [reply table/tables/index.go:184]: We can enable it at the beginning of `upgrade`.

Comment (coocood) [reply table/tables/index.go:184]: But we don't need to do it in this PR.

Comment (tangenta) [reply table/tables/index.go:184]: If the `upgrade()` changes the schema of `mysql.tidb`, we will encounter the same problem.

**Pull request merged**